File: backend/src/main/scala/services/governance/objects/GovernanceTypes.scala
package services.governance.objects

import java.util.Locale

import services.battle.objects.EpochMillis
import services.mail.objects.{GovernanceMailMetadata, MailImportance, MailKind, MailReadState}

final case class AdminHandle(value: String) extends AnyVal {
  def key: String = value.toLowerCase(Locale.ROOT)
}

object AdminHandle {
  def fromString(value: String): Option[AdminHandle] = {
    val trimmed = Option(value).getOrElse("").trim
    Option.when(trimmed.equalsIgnoreCase("admin"))(AdminHandle(trimmed))
  }
}

final case class GovernanceActorHandle(value: String) extends AnyVal
final case class GovernanceTargetHandle(value: String) extends AnyVal {
  def key: String = value.toLowerCase(Locale.ROOT)
}

final case class ContributionAdjustmentId(value: String) extends AnyVal
final case class ContributionDelta(value: Int) extends AnyVal
final case class GovernanceReason(value: String) extends AnyVal
final case class GovernanceSourceLabel(value: String) extends AnyVal
final case class GovernanceSourcePath(value: String) extends AnyVal

final case class GovernanceReviewNotificationId(value: String) extends AnyVal
final case class GovernanceReviewTargetId(value: String) extends AnyVal
final case class GovernanceReviewTargetTitle(value: String) extends AnyVal
final case class GovernanceReviewTargetPath(value: String) extends AnyVal
final case class GovernanceReviewBody(value: String) extends AnyVal

final case class GovernanceMailSnapshotId(value: String) extends AnyVal

enum GovernanceReviewKind {
  case ReplayProposal
  case ReplayReport
  case DiscussionReport
  case BotSuggestion
}

object GovernanceReviewKind {
  def fromWire(value: String): Option[GovernanceReviewKind] =
    Option(value).map(_.trim) match {
      case Some("replay_proposal")   => Some(GovernanceReviewKind.ReplayProposal)
      case Some("replay_report")     => Some(GovernanceReviewKind.ReplayReport)
      case Some("discussion_report") => Some(GovernanceReviewKind.DiscussionReport)
      case Some("bot_suggestion")    => Some(GovernanceReviewKind.BotSuggestion)
      case _                         => None
    }

  def wireValue(value: GovernanceReviewKind): String =
    value match {
      case GovernanceReviewKind.ReplayProposal  => "replay_proposal"
      case GovernanceReviewKind.ReplayReport    => "replay_report"
      case GovernanceReviewKind.DiscussionReport => "discussion_report"
      case GovernanceReviewKind.BotSuggestion   => "bot_suggestion"
    }

  def displayLabel(value: GovernanceReviewKind): String =
    value match {
      case GovernanceReviewKind.ReplayProposal  => "Replay proposal"
      case GovernanceReviewKind.ReplayReport    => "Replay report"
      case GovernanceReviewKind.DiscussionReport => "Discussion report"
      case GovernanceReviewKind.BotSuggestion   => "Bot suggestion"
    }
}

enum GovernanceReviewTargetType {
  case Replay
  case Discussion
  case Bot
}

object GovernanceReviewTargetType {
  def fromWire(value: String): Option[GovernanceReviewTargetType] =
    Option(value).map(_.trim) match {
      case Some("replay")     => Some(GovernanceReviewTargetType.Replay)
      case Some("discussion") => Some(GovernanceReviewTargetType.Discussion)
      case Some("bot")        => Some(GovernanceReviewTargetType.Bot)
      case _                  => None
    }

  def wireValue(value: GovernanceReviewTargetType): String =
    value match {
      case GovernanceReviewTargetType.Replay     => "replay"
      case GovernanceReviewTargetType.Discussion => "discussion"
      case GovernanceReviewTargetType.Bot        => "bot"
    }
}

final case class ContributionAdjustmentRecord(
  id: ContributionAdjustmentId,
  actorHandle: AdminHandle,
  targetHandle: GovernanceTargetHandle,
  delta: ContributionDelta,
  reason: GovernanceReason,
  createdAt: EpochMillis,
  sourceLabel: GovernanceSourceLabel,
  sourcePath: GovernanceSourcePath
)

final case class GovernanceReviewNotificationRecord(
  id: GovernanceReviewNotificationId,
  actorHandle: GovernanceActorHandle,
  kind: GovernanceReviewKind,
  targetType: GovernanceReviewTargetType,
  targetId: GovernanceReviewTargetId,
  targetTitle: GovernanceReviewTargetTitle,
  targetPath: GovernanceReviewTargetPath,
  body: GovernanceReviewBody,
  createdAt: EpochMillis,
  mailId: GovernanceMailSnapshotId
)

final case class GovernanceMailSnapshot(
  id: GovernanceMailSnapshotId,
  ownerHandle: GovernanceTargetHandle,
  kind: MailKind,
  subject: String,
  excerpt: String,
  senderLabel: String,
  readState: MailReadState,
  importance: MailImportance,
  createdAt: EpochMillis,
  governanceMetadata: Option[GovernanceMailMetadata]
) {
  def unread: Boolean =
    MailReadState.unreadFlag(readState)

  def important: Boolean =
    MailImportance.importantFlag(importance)
}

File: frontend/src/objects/governance/governanceTypes.ts
import type { MailImportanceDto, MailKindDto, MailReadStateDto } from "../mail/mailTypes";

export type AdminHandle = string;
export type GovernanceActorHandle = string;
export type GovernanceTargetHandle = string;
export type ContributionAdjustmentId = string;
export type ContributionDelta = number;
export type GovernanceReason = string;
export type GovernanceSourceLabel = string;
export type GovernanceSourcePath = string;
export type GovernanceReviewNotificationId = string;
export type GovernanceReviewTargetId = string;
export type GovernanceReviewTargetTitle = string;
export type GovernanceReviewTargetPath = string;
export type GovernanceReviewBody = string;
export type GovernanceMailSnapshotId = string;

export type GovernanceReviewKindDto =
  | "replay_proposal"
  | "replay_report"
  | "discussion_report"
  | "bot_suggestion";

export type GovernanceReviewTargetTypeDto = "replay" | "discussion" | "bot";

export interface ContributionAdjustmentRecord {
  id: ContributionAdjustmentId;
  actorHandle: AdminHandle;
  targetHandle: GovernanceTargetHandle;
  delta: ContributionDelta;
  reason: GovernanceReason;
  createdAt: number;
  sourceLabel: GovernanceSourceLabel;
  sourcePath: GovernanceSourcePath;
}

export interface GovernanceReviewNotificationRecord {
  id: GovernanceReviewNotificationId;
  actorHandle: GovernanceActorHandle;
  kind: GovernanceReviewKindDto;
  targetType: GovernanceReviewTargetTypeDto;
  targetId: GovernanceReviewTargetId;
  targetTitle: GovernanceReviewTargetTitle;
  targetPath: GovernanceReviewTargetPath;
  body: GovernanceReviewBody;
  createdAt: number;
  mailId: GovernanceMailSnapshotId;
}

export interface GovernanceMailSnapshot {
  id: GovernanceMailSnapshotId;
  ownerHandle: GovernanceTargetHandle;
  kind: MailKindDto;
  subject: string;
  excerpt: string;
  senderLabel: string;
  readState: MailReadStateDto;
  importance: MailImportanceDto;
  createdAt: number;
  governanceMetadata?: GovernanceMailMetadata;
}

export interface GovernanceMailMetadata {
  actorHandle: GovernanceActorHandle;
  targetPath: GovernanceReviewTargetPath;
  targetLabel: GovernanceReviewTargetTitle;
}

export interface ContributionAdjustmentApiRequestDto {
  actorHandle: string;
  targetHandle: string;
  delta: number;
  reason: string;
  sourceLabel: string;
  sourcePath: string;
}

export interface GovernanceReviewNotificationApiRequestDto {
  actorHandle: string;
  kind: string;
  targetType: string;
  targetId: string;
  targetTitle: string;
  targetPath: string;
  body: string;
}

export interface ContributionAdjustmentItemResponseDto {
  id: string;
  actorHandle: string;
  targetHandle: string;
  delta: number;
  reason: string;
  createdAt: number;
  sourceLabel: string;
  sourcePath: string;
}

export interface GovernanceReviewNotificationItemResponseDto {
  id: string;
  actorHandle: string;
  kind: GovernanceReviewKindDto;
  targetType: GovernanceReviewTargetTypeDto;
  targetId: string;
  targetTitle: string;
  targetPath: string;
  body: string;
  createdAt: number;
  mailId: string;
}

export interface GovernanceMailSnapshotResponseDto {
  id: string;
  ownerHandle: string;
  kind: MailKindDto;
  subject: string;
  excerpt: string;
  senderLabel: string;
  unread: boolean;
  important: boolean;
  createdAt: number;
  governanceActorHandle?: string;
  governanceTargetPath?: string;
  governanceTargetLabel?: string;
}

export interface ContributionAdjustmentListResponseDto {
  adjustments: ContributionAdjustmentItemResponseDto[];
}

export interface ContributionAdjustmentCreateResponseDto {
  ok: boolean;
  adjustment: ContributionAdjustmentItemResponseDto;
  mail: GovernanceMailSnapshotResponseDto;
}

export interface GovernanceReviewNotificationListResponseDto {
  notifications: GovernanceReviewNotificationItemResponseDto[];
}

export interface GovernanceReviewNotificationCreateResponseDto {
  ok: boolean;
  notification: GovernanceReviewNotificationItemResponseDto;
  mail: GovernanceMailSnapshotResponseDto;
}
